sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -510,6 +510,15 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case EqualTo(c @ CaseWhen(branches, elseValue), right)
+          if c.deterministic &&
```

Comment:
More precisely, I think we only need to make sure the skipped branches are all deterministic.
```
val (picked, skipped) = branches.partition(_._2.equals(right))
if (skipped.forall(_._1.determinisitc)) {
  ...
} else {
  original
}
```